File: mm-ts/src/mm/binder.ts
import { MMValue, MMObject, MMArray } from '../ast/ast';
import { Tag } from '../ast/tag';
import { Node } from '../ast/ast';
import { ValueType } from '../ast/value-type';

export type Constructor<T> = new () => T;

export class Binder {
  bind<T>(node: Node, type: Constructor<T> | T): T {
    // const result = new Type();
    const result =
      typeof type === 'function' ? new (type as Constructor<T>)() : type;
    // console.log('result', result);
    const { error } = this.bindNode(node, result);
    if (error) {
      throw new Error(error);
    }
    return result;
  }

  bindNode(node: Node, target: any): { value: any; error: string | null } {
    if (node instanceof MMValue) {
      return this.convertValue(node, target);
    }

    if (node instanceof MMObject) {
      const tag = node.getTag();
      if (tag?.type === ValueType.Object) {
        return this.convertStruct(node, target);
      } else {
        return this.convertMap(node, target);
      }
    }

    if (node instanceof MMArray) {
      const tag = node.getTag();
      if (tag?.type === ValueType.Array) {
        return this.convertArray(node, target);
      } else {
        return this.convertSlice(node, target);
      }
    }

    return { value: target, error: `unsupported node type: ${typeof node}` };
  }

  private convertStruct(
    obj: MMObject,
    target: any,
  ): { value: any; error: string | null } {
    const tag = obj.getTag();
    if (tag?.nullable && target === null) {
      return {
        value: target,
        error: 'convertStruct requires object type, got null',
      };
    }
    if (typeof target !== 'object' || target === null) {
      return {
        value: target,
        error: `convertStruct requires object type, got ${typeof target}`,
      };
    }

    const properties = obj.getProperties();
    for (const [key, valueNode] of Object.entries(properties)) {
      const camelKey = key.replace(/_([a-z])/g, (_, c) => c.toUpperCase());
      if (!(camelKey in target)) {
        return { value: target, error: `struct has no field '${camelKey}'` };
      }

      // 🔥 拿到子节点的值，直接赋值！
      const { value, error } = this.bindNode(valueNode, target[camelKey]);
      if (error) {
        return {
          value: target,
          error: `failed to bind field ${camelKey}: ${error}`,
        };
      }

      // ✅ 赋值生效！解决指针问题！
      target[camelKey] = value;
    }

    return { value: target, error: null };
  }

  private convertMap(
    obj: MMObject,
    target: any,
  ): { value: any; error: string | null } {
    if (
      typeof target !== 'object' ||
      target === null ||
      Array.isArray(target)
    ) {
      return { value: target, error: `convertMap requires object` };
    }

    const properties = obj.getProperties();
    for (const [key, valueNode] of Object.entries(properties)) {
      const { value, error } = this.bindNode(valueNode, target[key]);
      if (error) {
        return { value: target, error: `field ${key}: ${error}` };
      }
      target[key] = value;
    }
    return { value: target, error: null };
  }

  private convertArray(
    arr: MMArray,
    target: any,
  ): { value: any; error: string | null } {
    if (!Array.isArray(target)) {
      return { value: target, error: `convertArray requires array` };
    }

    const elements = arr.getElements();
    const size = arr.getTag()?.size || elements.length;

    if (target.length !== size) {
      return {
        value: target,
        error: `array length mismatch: expected ${size}, got ${target.length}`,
      };
    }

    for (let i = 0; i < elements.length; i++) {
      const el = elements[i];
      if (!el) continue;

      const { value, error } = this.bindNode(el, target[i]);
      if (error) {
        return { value: target, error: `array[${i}]: ${error}` };
      }
      target[i] = value;
    }

    return { value: target, error: null };
  }

  private convertSlice(
    arr: MMArray,
    target: any[],
  ): { value: any; error: string | null } {
    if (!Array.isArray(target)) {
      return { value: target, error: `convertSlice requires array` };
    }

    const elements = arr.getElements();
    target.length = 0;

    for (const el of elements) {
      const defaultValue = this.createDefaultValue(el.getTag());
      const { value, error } = this.bindNode(el, defaultValue);
      if (error) {
        return { value: target, error };
      }
      target.push(value);
    }

    return { value: target, error: null };
  }

  private createDefaultValue(tag: Tag | null): any {
    if (!tag) return null;

    switch (tag.type) {
      case ValueType.Int:
      case ValueType.Int8:
      case ValueType.Int16:
      case ValueType.Int32:
      case ValueType.Int64:
        return 0;
      case ValueType.Uint:
      case ValueType.Uint8:
      case ValueType.Uint16:
      case ValueType.Uint32:
      case ValueType.Uint64:
        return 0;
      case ValueType.Float32:
      case ValueType.Float64:
        return 0.0;
      case ValueType.String:
      case ValueType.Email:
      case ValueType.URL:
      case ValueType.IP:
      case ValueType.UUID:
      case ValueType.Decimal:
        return '';
      case ValueType.Bool:
        return false;
      case ValueType.Bytes:
        return new Uint8Array();
      case ValueType.BigInt:
        return BigInt(0);
      case ValueType.DateTime:
      case ValueType.Date:
      case ValueType.Time:
        return new Date(0);
      case ValueType.Enum:
        return '';
      default:
        return null;
    }
  }

  private convertValue(
    val: MMValue,
    target: any,
  ): { value: any; error: string | null } {
    const tag = val.getTag();
    if (!tag) return { value: target, error: 'tag is null' };

    const data = val.getValue();

    // 直接返回新值，不赋值！
    switch (tag.type) {
      case ValueType.Int:
      case ValueType.Int8:
      case ValueType.Int16:
      case ValueType.Int32:
      case ValueType.Int64:
      case ValueType.Uint:
      case ValueType.Uint8:
      case ValueType.Uint16:
      case ValueType.Uint32:
      case ValueType.Uint64:
        return typeof data === 'bigint'
          ? { value: Number(data), error: null }
          : { value: data, error: null };

      case ValueType.Float32:
      case ValueType.Float64:
        return { value: Number(data), error: null };

      case ValueType.String:
      case ValueType.Email:
      case ValueType.URL:
      case ValueType.IP:
      case ValueType.Decimal:
      case ValueType.Enum:
        return { value: String(data), error: null };

      case ValueType.Bool:
        return { value: Boolean(data), error: null };

      case ValueType.Bytes:
        return { value: new Uint8Array(data), error: null };

      case ValueType.BigInt:
        return { value: BigInt(data), error: null };

      case ValueType.DateTime:
      case ValueType.Date:
      case ValueType.Time:
        return { value: new Date(data), error: null };

      case ValueType.UUID:
        if (data instanceof Uint8Array) {
          const s = Array.from(data)
            .map((b) => b.toString(16).padStart(2, '0'))
            .join('');
          const uuid = `${s.slice(0, 8)}-${s.slice(8, 12)}-${s.slice(12, 16)}-${s.slice(16, 20)}-${s.slice(20)}`;
          return { value: uuid, error: null };
        }
        return { value: String(data), error: null };
    }

    return { value: target, error: `unsupported value type: ${tag.type}` };
  }
}

export function bindJSONC<T>(node: Node, Type: new () => T): T {
  const binder = new Binder();
  return binder.bind(node, Type);
}

File: internal/core/binder.go
package core

import (
	"fmt"
	"math/big"
	"net"
	"net/url"
	"reflect"
	"time"
	"unicode"

	"github.com/metamessage/metamessage/internal/ir"
)

func Bind(node ir.Node, out any) error {
	outVal := reflect.ValueOf(out)
	if outVal.Kind() != reflect.Pointer || outVal.IsNil() {
		return fmt.Errorf("out must be a non-nil pointer")
	}

	outVal = outVal.Elem()

	switch n := node.(type) {
	case *ir.Object:
		if n.Tag.Type == ir.ValueTypeObj {
			return convertStruct(n, outVal)
		} else {
			return convertMap(n, outVal)
		}

	case *ir.Array:
		if n.Tag.Type == ir.ValueTypeArr {
			return convertArray(n, outVal)
		} else {
			return convertSlice(n, outVal)
		}

	case *ir.Value:
		return convertValue(n, outVal)

	default:
		return fmt.Errorf("unsupported node type: %T", node)
	}
}

func convertStruct(obj *ir.Object, outVal reflect.Value) error {
	if obj.Tag.Nullable {
		if outVal.Kind() != reflect.Pointer {
			return fmt.Errorf("convertStruct requires pointer type, got %s", outVal.Kind())
		}
		if outVal.IsNil() {
			return fmt.Errorf("convertStruct: nullable pointer must point to struct, got nil")
		}
		outVal = outVal.Elem()
	}
	if outVal.Kind() != reflect.Struct {
		return fmt.Errorf("convertStruct requires struct type, got %s", outVal.Kind())
	}

	for _, field := range obj.Fields {
		fieldKey := field.Key
		var fieldVal reflect.Value
		found := false
		runes := []rune(fieldKey)
		runes[0] = unicode.ToUpper(runes[0])
		name := string(runes)
		for i := 0; i < outVal.NumField(); i++ {
			structField := outVal.Type().Field(i)
			if structField.Name == name {
				fieldVal = outVal.Field(i)
				found = true
				break
			}
		}

		if !found {
			return fmt.Errorf("struct has no field '%s'", fieldKey)
		}

		if !fieldVal.CanSet() {
			return fmt.Errorf("struct field '%s' is unexported (cannot set)", fieldKey)
		}

		if err := Bind(field.Value, fieldVal.Addr().Interface()); err != nil {
			return fmt.Errorf("failed to bind struct field %s: %w", field.Key, err)
		}
	}

	return nil
}

func convertMap(obj *ir.Object, outVal reflect.Value) error {
	if obj.Tag.Nullable {
		if outVal.Kind() != reflect.Pointer {
			return fmt.Errorf("convertMap requires pointer type, got %s", outVal.Kind())
		}
		if outVal.IsNil() {
			return fmt.Errorf("convertMap: nullable pointer must point to map, got nil")
		}
		outVal = outVal.Elem()
	}
	if outVal.Kind() != reflect.Map {
		return fmt.Errorf("convertMap requires map type, got %s", outVal.Kind())
	}

	if outVal.IsNil() {
		outVal.Set(reflect.MakeMap(outVal.Type()))
	}

	if outVal.Type().Key().Kind() != reflect.String {
		return fmt.Errorf("convertMap requires map with string key, got %s", outVal.Type().Key().Kind())
	}

	if outVal.Len() > 0 {
		return fmt.Errorf("target map is not empty (length: %d), clear it before binding", outVal.Len())
	}

	mapValType := outVal.Type().Elem()
	for _, field := range obj.Fields {
		key := reflect.ValueOf(field.Key)
		val := reflect.New(mapValType).Elem()
		if err := Bind(field.Value, val.Addr().Interface()); err != nil {
			return fmt.Errorf("failed to convert field %s: %w", field.Key, err)
		}

		outVal.SetMapIndex(key, val)
	}
	return nil
}

func convertArray(arr *ir.Array, outVal reflect.Value) error {
	if arr.Tag.Nullable {
		if outVal.Kind() != reflect.Pointer {
			return fmt.Errorf("convert array requires pointer type, got %s", outVal.Kind())
		}
		if outVal.IsNil() {
			return fmt.Errorf("convert array: nullable pointer must point to array, got nil")
		}
		outVal = outVal.Elem()
	}
	if outVal.Kind() != reflect.Array {
		return fmt.Errorf("convert array requires array type, got %s", outVal.Kind())
	}

	arrayLen := outVal.Len()
	size := arr.Tag.Size
	if size != arrayLen {
		return fmt.Errorf("array length mismatch: target array length %d, got %d items", arrayLen, size)
	}

	for i, item := range arr.Items {
		elem := outVal.Index(i)
		if err := Bind(item, elem.Addr().Interface()); err != nil {
			return fmt.Errorf("failed to convert array item %d: %w", i, err)
		}
	}

	return nil
}

func convertSlice(arr *ir.Array, outVal reflect.Value) error {
	if arr.Tag.Nullable {
		if outVal.Kind() != reflect.Pointer {
			return fmt.Errorf("convert slice requires pointer type, got %s", outVal.Kind())
		}
		if outVal.IsNil() {
			return fmt.Errorf("convert slice: nullable pointer must point to slice, got nil")
		}
		outVal = outVal.Elem()
	}
	if outVal.Kind() != reflect.Slice {
		return fmt.Errorf("convert slice requires slice type, got %s", outVal.Kind())
	}

	size := len(arr.Items)
	slice := reflect.MakeSlice(outVal.Type(), size, size)

	for i, item := range arr.Items {
		elem := slice.Index(i)
		if err := Bind(item, elem.Addr().Interface()); err != nil {
			return fmt.Errorf("failed to convert slice item %d: %w", i, err)
		}
	}

	outVal.Set(slice)
	return nil
}

func convertValue(val *ir.Value, outVal reflect.Value) error {
	if val.Tag.Nullable {
		if outVal.Kind() != reflect.Pointer {
			return fmt.Errorf("convertValue requires pointer type, got %s", outVal.Kind())
		}
		if outVal.IsNil() {
			outVal.Set(reflect.New(outVal.Type().Elem()))
		}
		outVal = outVal.Elem()
	}

	data := val.Data
	text := val.Text

	tag := val.GetTag()
	if tag == nil {
		return fmt.Errorf("")
	}

	switch tag.Type {
	case ir.ValueTypeTime, ir.ValueTypeDate, ir.ValueTypeDatetime:
		targetType := reflect.TypeFor[time.Time]()
		if outVal.Type() != targetType {
			return fmt.Errorf("target type must be time.Time, got %s", outVal.Type())
		}
		if !outVal.CanAddr() {
			return fmt.Errorf("unexpected: target time.Time value is unaddressable (bug)")
		}
		switch d := data.(type) {
		case time.Time:
			*(*time.Time)(outVal.Addr().UnsafePointer()) = d
		default:
			return fmt.Errorf("unsupported type for time conversion: %T (expected time.Time)", d)
		}

	case ir.ValueTypeBigint:
		targetType := reflect.TypeFor[big.Int]()
		if outVal.Type() != targetType {
			return fmt.Errorf("target type must be big.Int, got %s", outVal.Type())
		}

		switch d := data.(type) {
		case big.Int:
			*(*big.Int)(outVal.UnsafePointer()) = d
		default:
			return fmt.Errorf("unsupported type for big.Int conversion: %T", d)
		}

	case ir.ValueTypeUuid:
		targetType := reflect.TypeFor[string]()
		if outVal.Type() != targetType {
			return fmt.Errorf("target type must be string, got %s", outVal.Type())
		}

		switch d := data.(type) {
		case string:
			*(*string)(outVal.UnsafePointer()) = d
		default:
			return fmt.Errorf("unsupported type for uuid conversion: %T", d)
		}

	case ir.ValueTypeDecimal:
		targetType := reflect.TypeFor[string]()
		if outVal.Type() != targetType {
			return fmt.Errorf("target type must be string, got %s", outVal.Type())
		}

		switch d := data.(type) {
		case string:
			*(*string)(outVal.UnsafePointer()) = d
		default:
			return fmt.Errorf("unsupported type for decimal conversion: %T", d)
		}

	case ir.ValueTypeEmail:
		targetType := reflect.TypeFor[string]()
		if outVal.Type() != targetType {
			return fmt.Errorf("target type must be string, got %s", outVal.Type())
		}

		switch d := data.(type) {
		case string:
			*(*string)(outVal.UnsafePointer()) = d
		default:
			return fmt.Errorf("unsupported type for email conversion: %T", d)
		}

	case ir.ValueTypeIp:
		targetType := reflect.TypeFor[net.IP]()
		if outVal.Type() != targetType {
			return fmt.Errorf("target type must be net.IP, got %s", outVal.Type())
		}

		switch d := data.(type) {
		case net.IP:
			*(*net.IP)(outVal.UnsafePointer()) = d
		default:
			return fmt.Errorf("unsupported type for ip conversion: %T", d)
		}

	case ir.ValueTypeUrl:
		targetType := reflect.TypeFor[url.URL]()
		if outVal.Type() != targetType {
			return fmt.Errorf("target type must be url.URL, got %s", outVal.Type())
		}

		switch d := data.(type) {
		case url.URL:
			*(*url.URL)(outVal.UnsafePointer()) = d
		default:
			return fmt.Errorf("unsupported type for url conversion: %T", d)
		}

	case ir.ValueTypeEnum:
		if outVal.Type() != reflect.TypeFor[string]() {
			return fmt.Errorf("target type must be string, got %s", outVal.Type())
		}
		if !outVal.CanAddr() {
			return fmt.Errorf("unexpected: target string value is unaddressable (bug)")
		}
		*(*string)(outVal.Addr().UnsafePointer()) = text

	case ir.ValueTypeImage:
		return fmt.Errorf("unsupported type: %s", tag.Type)
	case ir.ValueTypeVideo:
		return fmt.Errorf("unsupported type: %s", tag.Type)
	case ir.ValueTypeDoc:
		return fmt.Errorf("unsupported type: %s", tag.Type)
	// case ir.ValueTypeAudio:
	// 	return fmt.Errorf("unsupported type: %s", tag.Type)
	// case ir.ValueTypeFont:
	// 	return fmt.Errorf("unsupported type: %s", tag.Type)

	case ir.ValueTypeI:
		if outVal.Type() != reflect.TypeFor[int]() {
			return fmt.Errorf("target type must be int, got %s", outVal.Type())
		}
		if !outVal.CanAddr() {
			return fmt.Errorf("unexpected: target int value is unaddressable (bug)")
		}
		switch d := data.(type) {
		case int:
			*(*int)(outVal.Addr().UnsafePointer()) = d
		default:
			return fmt.Errorf("unsupported type for int conversion: %T", d)
		}

	case ir.ValueTypeI8:
		if outVal.Type() != reflect.TypeFor[int8]() {
			return fmt.Errorf("target type must be int8, got %s", outVal.Type())
		}
		if !outVal.CanAddr() {
			return fmt.Errorf("unexpected: target int8 value is unaddressable (bug)")
		}
		switch d := data.(type) {
		case int8:
			*(*int8)(outVal.Addr().UnsafePointer()) = d
		default:
			return fmt.Errorf("unsupported type for int8 conversion: %T", d)
		}

	case ir.ValueTypeI16:
		if outVal.Type() != reflect.TypeFor[int16]() {
			return fmt.Errorf("target type must be int16, got %s", outVal.Type())
		}
		if !outVal.CanAddr() {
			return fmt.Errorf("unexpected: target int16 value is unaddressable (bug)")
		}
		switch d := data.(type) {
		case int16:
			*(*int16)(outVal.Addr().UnsafePointer()) = d
		default:
			return fmt.Errorf("unsupported type for int16 conversion: %T (expected int16)", d)
		}

	case ir.ValueTypeI32:
		if outVal.Type() != reflect.TypeFor[int32]() {
			return fmt.Errorf("target type must be int32, got %s", outVal.Type())
		}
		if !outVal.CanAddr() {
			return fmt.Errorf("unexpected: target int32 value is unaddressable (bug)")
		}
		switch d := data.(type) {
		case int32:
			*(*int32)(outVal.Addr().UnsafePointer()) = d
		default:
			return fmt.Errorf("unsupported type for int32 conversion: %T (expected int32)", d)
		}

	case ir.ValueTypeI64:
		if outVal.Type() != reflect.TypeFor[int64]() {
			return fmt.Errorf("target type must be int64, got %s", outVal.Type())
		}
		if !outVal.CanAddr() {
			return fmt.Errorf("unexpected: target int64 value is unaddressable (bug)")
		}
		switch d := data.(type) {
		case int64:
			*(*int64)(outVal.Addr().UnsafePointer()) = d
		default:
			return fmt.Errorf("unsupported type for int64 conversion: %T (expected int64)", d)
		}

	case ir.ValueTypeU:
		if outVal.Type() != reflect.TypeFor[uint]() {
			return fmt.Errorf("target type must be uint, got %s", outVal.Type())
		}
		if !outVal.CanAddr() {
			return fmt.Errorf("unexpected: target uint value is unaddressable (bug)")
		}
		switch d := data.(type) {
		case uint:
			*(*uint)(outVal.Addr().UnsafePointer()) = d
		default:
			return fmt.Errorf("unsupported type for uint conversion: %T (expected uint)", d)
		}

	case ir.ValueTypeU8:
		if outVal.Type() != reflect.TypeFor[uint8]() {
			return fmt.Errorf("target type must be uint8, got %s", outVal.Type())
		}
		if !outVal.CanAddr() {
			return fmt.Errorf("unexpected: target uint8 value is unaddressable (bug)")
		}
		switch d := data.(type) {
		case uint8:
			*(*uint8)(outVal.Addr().UnsafePointer()) = d
		default:
			return fmt.Errorf("unsupported type for uint8 conversion: %T (expected uint8)", d)
		}

	case ir.ValueTypeU16:
		if outVal.Type() != reflect.TypeFor[uint16]() {
			return fmt.Errorf("target type must be uint16, got %s", outVal.Type())
		}
		if !outVal.CanAddr() {
			return fmt.Errorf("unexpected: target uint16 value is unaddressable (bug)")
		}
		switch d := data.(type) {
		case uint16:
			*(*uint16)(outVal.Addr().UnsafePointer()) = d
		default:
			return fmt.Errorf("unsupported type for uint16 conversion: %T (expected uint16)", d)
		}

	case ir.ValueTypeU32:
		if outVal.Type() != reflect.TypeFor[uint32]() {
			return fmt.Errorf("target type must be uint32, got %s", outVal.Type())
		}
		if !outVal.CanAddr() {
			return fmt.Errorf("unexpected: target uint32 value is unaddressable (bug)")
		}
		switch d := data.(type) {
		case uint32:
			*(*uint32)(outVal.Addr().UnsafePointer()) = d
		default:
			return fmt.Errorf("unsupported type for uint32 conversion: %T (expected uint32)", d)
		}

	case ir.ValueTypeU64:
		if outVal.Type() != reflect.TypeFor[uint64]() {
			return fmt.Errorf("target type must be uint64, got %s", outVal.Type())
		}
		if !outVal.CanAddr() {
			return fmt.Errorf("unexpected: target uint64 value is unaddressable (bug)")
		}
		switch d := data.(type) {
		case uint64:
			*(*uint64)(outVal.Addr().UnsafePointer()) = d
		default:
			return fmt.Errorf("unsupported type for uint64 conversion: %T (expected uint64)", d)
		}

	case ir.ValueTypeF32:
		if outVal.Type() != reflect.TypeFor[float32]() {
			return fmt.Errorf("target type must be float32, got %s", outVal.Type())
		}
		if !outVal.CanAddr() {
			return fmt.Errorf("unexpected: target float32 value is unaddressable (bug)")
		}
		switch d := data.(type) {
		case float32:
			*(*float32)(outVal.Addr().UnsafePointer()) = d
		default:
			return fmt.Errorf("unsupported type for float32 conversion: %T (expected float32)", d)
		}

	case ir.ValueTypeF64:
		if outVal.Type() != reflect.TypeFor[float64]() {
			return fmt.Errorf("target type must be float64, got %s", outVal.Type())
		}
		if !outVal.CanAddr() {
			return fmt.Errorf("unexpected: target float64 value is unaddressable (bug)")
		}
		switch d := data.(type) {
		case float64:
			*(*float64)(outVal.Addr().UnsafePointer()) = d
		default:
			return fmt.Errorf("unsupported type for float64 conversion: %T (expected float64)", d)
		}

	case ir.ValueTypeStr:
		if outVal.Type() != reflect.TypeFor[string]() {
			return fmt.Errorf("target type must be string, got %s", outVal.Type())
		}
		if !outVal.CanAddr() {
			return fmt.Errorf("unexpected: target string value is unaddressable (bug)")
		}
		*(*string)(outVal.Addr().UnsafePointer()) = text

	case ir.ValueTypeBool:
		if outVal.Type() != reflect.TypeFor[bool]() {
			return fmt.Errorf("target type must be bool, got %s", outVal.Type())
		}
		if !outVal.CanAddr() {
			return fmt.Errorf("unexpected: target bool value is unaddressable (bug)")
		}
		switch d := data.(type) {
		case bool:
			*(*bool)(outVal.Addr().UnsafePointer()) = d
		default:
			return fmt.Errorf("unsupported type for bool conversion: %T (expected bool)", d)
		}

	case ir.ValueTypeBytes:
		if outVal.Type() != reflect.TypeFor[[]byte]() {
			return fmt.Errorf("target type must be []byte, got %s", outVal.Type())
		}
		if !outVal.CanAddr() {
			return fmt.Errorf("unexpected: target []byte value is unaddressable (bug)")
		}
		switch d := data.(type) {
		case []byte:
			*(*[]byte)(outVal.Addr().UnsafePointer()) = d
		default:
			return fmt.Errorf("unsupported type for bytes conversion: %T (expected []byte)", d)
		}

	case ir.ValueTypeMap:
		return fmt.Errorf("unsupported type: %s", tag.Type)
	case ir.ValueTypeObj:
		return fmt.Errorf("unsupported type: %s", tag.Type)
	case ir.ValueTypeArr:
		return fmt.Errorf("unsupported type: %s", tag.Type)
	case ir.ValueTypeVec:
		return fmt.Errorf("unsupported type: %s", tag.Type)
	case ir.ValueTypeUnknown:
		return fmt.Errorf("unsupported type: %s", tag.Type)
	default:
		return fmt.Errorf("unsupported type: %s", tag.Type)
	}

	return nil
}
